Library/Homebrew/keg_relocate.rb

```diff
@@ -163,6 +167,45 @@ def replace_text_in_files(relocation, files: nil)
     changed_files
   end
 
+  def replace_build_prefix_in_file(old_cellar, new_prefix, new_cellar, file)
+    old_prefix = Pathname(old_cellar).parent.to_s
+
+    old_prefix_optdir = "#{old_prefix}/opt"
+    old_prefix_etcdir = "#{old_prefix}/etc"
+    old_prefix_vardir = "#{old_prefix}/var"
+    old_prefix_vim_sharedir = "#{old_prefix}/share/vim"
+
+    new_prefix_optdir = "#{new_prefix}/opt"
+    new_prefix_etcdir = "#{new_prefix}/etc"
+    new_prefix_vardir = "#{new_prefix}/var"
+    new_prefix_vim_sharedir = "#{new_prefix}/share/vim"
+
+    old_prefix_dirs = [old_cellar, old_prefix_optdir, old_prefix_etcdir, old_prefix_vardir, old_prefix_vim_sharedir]
+    old_prefix_regex = Regexp.new(old_prefix_dirs.join("|"))
+
+    file.ensure_writable do
+      binary = File.binread file
+      # Skip patching if no string references to the cellar exist.
+      next if binary.scan(old_prefix_regex).empty?
+
+      # Split binary by null characters into array and substitute new cellar for old cellar.
+      # Null padding is added if the new string is too short.
+      patched_binary_strings = []
+      binary.split("\x00").each do |s|
+        patched_binary_strings << s.gsub(old_cellar, new_cellar)
+                                   .gsub(old_prefix_optdir, new_prefix_optdir)
+                                   .gsub(old_prefix_etcdir, new_prefix_etcdir)
+                                   .gsub(old_prefix_vardir, new_prefix_vardir)
+                                   .gsub(old_prefix_vim_sharedir, new_prefix_vim_sharedir)
+                                   .ljust(s.size, "\x00")
+      end
+
+      # Add back null padding at the end of the binary if needed.
+      patched_binary = patched_binary_strings.join("\x00").ljust(binary.size, "\x00")
```

Comment:
I don't quite follow why this is necessary.

Reply:
To expand on this comment, the reason for this is that in most binaries (at least all the ones I've seen so far), the last few characters are null bytes.  When we split by the null byte, the last few strings are empty, so when we join again at the end, the last character ends up being the last non-null byte.  To rectify this we need to add back those last null characters as padding.  If this is worth detailing further in the comment I am happy to do so!

Reply:
> If this is worth detailing further in the comment I am happy to do so!

Yes, please ❤️